compiler/generator_js.c

```diff
@@ -1278,39 +1278,71 @@ static void generate(struct generator * g, struct node * p) {
-        w(g, "// deno-lint-ignore-file~N"
-             "import ~P from './base-stemmer.mjs'~N"
-             "~N"
-             "/** @typedef {{ stemWord(word: string): string }} Stemmer */~N"
-             "~N"
-             "/** @type {{ new(): Stemmer }} */~N"
-             "~C ~n = function() {~+~N"
-             "~M~D base = new ~P();~N");
-    } else {
-        w(g, "/**@constructor*/~N"
-             "~D ~n = function() {~+~N"
-             "~M~C ~P = require('./base-stemmer.js');~N"
-             "~M~D base = new ~P();~N");
+    switch (g->options->js_type) {
+        case JS_GLOBAL:
+            w(g, "/**@constructor*/~N"
+                 "~D ~n = function() {~+~N"
+                 "~M~C ~P = require('./base-stemmer.js');~N"
+                 "~M~D base = new ~P();~N");
+            break;
+        case JS_ESM:
+            w(g, "// deno-lint-ignore-file~N"
+                 "import ~P from './base-stemmer.mjs'~N"
+                 "~N"
+                 "/** @typedef {{ stemWord(word: string): string }} Stemmer */~N"
+                 "~N"
+                 "/** @type {{ new(): Stemmer }} */~N"
+                 "~C ~n = function() {~+~N"
+                 "~M~D base = new ~P();~N");
+            break;
+        case JS_ES6:
+            w(g, "import ~P from './base-stemmer-es6.mjs'~N"
+                 "~N"
+                 "const ~n = (() => {~+~N"
+                 "~Mconst base = new ~P();~N");
+            break;
     }
     write_newline(g);
 }
 
 static void generate_class_end(struct generator * g) {
-    w(g, "~N");
-    w(g, "~M/**@return{string}*/~N");
-    w(g, "~Mthis['stemWord'] = function(/**string*/word) {~+~N");
-    w(g, "~Mbase.setCurrent(word);~N");
-    w(g, "~Mthis.stem();~N");
-    w(g, "~Mreturn base.getCurrent();~N");
-    w(g, "~-~M};~N");
-    w(g, "~-};~N");
-    if (g->options->js_esm) {
-        w(g, "~N"
-             "export default ~n~N");
-    } else {
-        w(g, "~N"
-             "if (typeof module === 'object' && module.exports) module.exports = ~n;~N");
+    write_newline(g);
+    switch (g->options->js_type) {
+        case JS_GLOBAL:
+            w(g, "~M/**@return{string}*/~N");
+            w(g, "~Mthis['stemWord'] = function(/**string*/word) {~+~N");
+            w(g, "~Mbase.setCurrent(word);~N");
+            w(g, "~Mthis.stem();~N");
+            w(g, "~Mreturn base.getCurrent();~N");
+            w(g, "~-~M};~N");
+            w(g, "~-};~N");
+            w(g, "~N"
+                 "if (typeof module === 'object' && module.exports) module.exports = ~n;~N");
+            break;
+        case JS_ESM:
+            w(g, "~M/**@return{string}*/~N");
+            w(g, "~Mthis['stemWord'] = function(/**string*/word) {~+~N");
+            w(g, "~Mbase.setCurrent(word);~N");
+            w(g, "~Mthis.stem();~N");
+            w(g, "~Mreturn base.getCurrent();~N");
+            w(g, "~-~M};~N");
+            w(g, "~-};~N");
+            w(g, "~N"
+                 "export default ~n~N");
```

Comment:
The merged version doesn't use `export default`.  The upshot of this seems to be that dynamically loading the module and creating an object now requires generating both the module name and class name from the language name, whereas with a default export you would only need to generate the module name.

Maybe we should provide a standard way to create a stemmer object from a language name (which could allow creating from an ISO-639 code too, like you can with C libstemmer).  It still seems a bit clumsy if that has to generate both a module name and a class name though.